File: 092_reverse_linked_list_ii/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def create_list(nums: list) -> ListNode:
    head = ListNode(nums[0])
    tmp = head
    for num in nums[1:]:
        tmp.next = ListNode(num)
        tmp = tmp.next

    return head

def print_list(head: ListNode) -> None:
    while head:
        print(head.val, end = ' ')
        head = head.next
    print()


class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if not head or not head.next or m >= n: return head

        dummy = ListNode(-1)
        dummy.next = head
        pre = dummy

        for i in range(1, m): pre = pre.next
        if not pre.next: return head

        node_before_m, node_m, cur = pre, pre.next, pre.next
        for i in range(m, n + 1):
            next_node = cur.next
            cur.next = pre
            pre = cur
            cur = next_node

        node_before_m.next = pre
        node_m.next = next_node

        return dummy.next


if __name__ == '__main__':
    head = create_list([1, 2, 3, 4, 5])
    result = Solution().reverseBetween(head, 2, 4)
    print_list(result)

File: 092_reverse_linked_list_ii/Solution.go
package main

import "fmt"

type ListNode struct {
    Val  int
    Next *ListNode
}

func createList(nums []int) *ListNode {
    head := &ListNode{nums[0], nil}
    tmp := head

    for _, num := range nums[1:] {
        tmp.Next = &ListNode{num, nil}
        tmp = tmp.Next
    }

    return head
}

func printList(head *ListNode) {
    for head != nil {
        fmt.Printf("%d ", head.Val)
        head = head.Next
    }
    fmt.Println()
}

func reverseBetween(head *ListNode, m int, n int) *ListNode {
    if head == nil || head.Next == nil || m >= n { return head }

    dummyHead := &ListNode{-1, head}
    pre := dummyHead

    for i := 1; i < m; i++ { pre = pre.Next }
    if pre.Next == nil { return head }

    nodeBeforeM, nodeM, cur, next := pre, pre.Next, pre.Next, pre
    for i := m; i <= n; i++ {
        next = cur.Next
        cur.Next = pre
        pre = cur
        cur = next
    }
    nodeBeforeM.Next = pre
    nodeM.Next = next

    return dummyHead.Next
}

func main() {
    head := createList([]int{1, 2, 3, 4, 5})
    result := reverseBetween(head, 2, 4)
    printList(result)
}
